media/client/main/source/ClientLogControl.cpp

```diff
@@ -0,0 +1,159 @@
+}
+} // namespace firebolt::rialto
+
+namespace firebolt::rialto::client
+{
+std::shared_ptr<IClientLogControlFactory> ClientLogControlFactory::createFactory()
+{
+    static std::shared_ptr<IClientLogControlFactory> factory;
+    if (!factory)
+    {
+        try
+        {
+            factory = std::make_shared<client::ClientLogControlFactory>();
+        }
+        catch (const std::exception &e)
+        {
+            RIALTO_CLIENT_LOG_ERROR("Failed to create ClientLogControlFactory, reason: %s", e.what());
+        }
+    }
+    return factory;
+}
+
+std::shared_ptr<IClientLogControl> ClientLogControlFactory::createClientLogControl()
+{
+    if (!m_singletonInstance)
+    {
+        try
+        {
+            m_singletonInstance = std::make_shared<ClientLogControl>();
+        }
+        catch (const std::exception &e)
+        {
+            RIALTO_CLIENT_LOG_ERROR("Failed to create ClientLogControl, reason: %s", e.what());
+        }
+    }
+    return m_singletonInstance;
+}
+
+ClientLogControl::ClientLogControl()
+{
+    RIALTO_CLIENT_LOG_DEBUG("entry:");
+}
+
+ClientLogControl::~ClientLogControl()
+{
+    RIALTO_CLIENT_LOG_DEBUG("entry:");
+    if (m_logHandler)
+        cancelLogHandler();
+}
+
+bool ClientLogControl::registerLogHandler(std::shared_ptr<IClientLogHandler> &handler, bool ignoreLogLevels)
+{
+    if (m_logHandler)
+    {
+        if (handler)
+        {
+            RIALTO_CLIENT_LOG_WARN("Replacing old log handler");
+        }
+        cancelLogHandler();
+    }
+
+    m_logHandler = handler;
+
+    if (!handler)
+        return true;
+
+    for (auto component : kAllComponentsToLog)
+    {
+        if (firebolt::rialto::logging::setLogHandler(component,
+                                                     std::bind(&ClientLogControl::forwardLog, this, component,
+                                                               std::placeholders::_1, std::placeholders::_2,
+                                                               std::placeholders::_3, std::placeholders::_4,
+                                                               std::placeholders::_5, std::placeholders::_6),
+                                                     ignoreLogLevels) !=
+            firebolt::rialto::logging::RIALTO_LOGGING_STATUS_OK)
+        {
+            RIALTO_CLIENT_LOG_WARN("Unable to register log handler");
+            cancelLogHandler();
+            return false;
+        }
+    }
+    return true;
+}
+
+void ClientLogControl::cancelLogHandler()
+{
+    RIALTO_CLIENT_LOG_INFO("Cancelling log handler");
+    for (auto component : kAllComponentsToLog)
+    {
+        firebolt::rialto::logging::setLogHandler(component, nullptr, false);
+    }
+    m_logHandler = nullptr;
+}
+
+void ClientLogControl::forwardLog(RIALTO_COMPONENT component, RIALTO_DEBUG_LEVEL level, const char *file, int line,
+                                  const char *function, const char *message, std::size_t messageLen)
+{
+    if (!m_logHandler)
```

Comment:
Don't want to put a mutex here, as it could unnecessarily slow logging down. Instead, I would take a copy of the shared_ptr at the beginning of the method so that it is not destroyed while this operation is ongoing.